src/util/ConfigManager/ConfigManager.cpp

```diff
@@ -34,24 +34,70 @@
+                                        std::string_view pathPrefix) {
+  std::vector<std::pair<std::string, ReturnPointer>> collectedOptions;
 
-    // Return a dereferenced reference.
-    return std::tie(pair.first, *pair.second);
-  });
+  std::ranges::for_each(
+      configurationOptions,
+      [&collectedOptions, &pathPrefix](auto pair) {
+        const std::string& pathToCurrentEntry =
+            absl::StrCat(pathPrefix, std::get<0>(pair));
+
+        std::visit(
+            [&collectedOptions, &pathToCurrentEntry]<typename T>(T& var) {
+              // A normal `ConfigOption` can be directly added. For a
+              // `ConfigManager` we have to recursively collect the options.
+              if constexpr (std::is_same_v<std::decay_t<T>, ConfigOption>) {
+                collectedOptions.push_back(
+                    std::make_pair(pathToCurrentEntry, &var));
+              } else {
+                AD_CORRECTNESS_CHECK(
+                    (std::is_same_v<std::decay_t<T>, ConfigManager>));
+                ad_utility::appendVector(
+                    collectedOptions,
+                    configurationOptionsImpl<ReturnPointer>(
+                        var.configurationOptions_, pathToCurrentEntry));
+              }
+            },
+            std::get<1>(pair));
+      },
+      [&pathPrefix](auto& pair) {
+        // Make sure, that there is no null pointer.
+        AD_CORRECTNESS_CHECK(pair.second != nullptr);
+
+        // An empty sub manager tends to point to a logic error on the user
+        // side.
+        if (const ConfigManager* ptr =
+                std::get_if<ConfigManager>(pair.second.get());
+            ptr != nullptr && ptr->configurationOptions_.empty()) {
+          throw std::runtime_error(absl::StrCat(
+              "The sub manager at '", pathPrefix, std::get<0>(pair),
+              "' is empty. Either fill it, or delete it."));
+        }
+
+        // Return a dereferenced reference.
+        return std::tie(pair.first, *pair.second);
+      });
+
+  return collectedOptions;
 }
 
 // ____________________________________________________________________________
-auto ConfigManager::configurationOptions() {
-  return configurationOptionsImpl(configurationOptions_);
+std::vector<std::pair<std::string, ConfigOption*>>
+ConfigManager::configurationOptions(std::string_view pathPrefix) {
+  return configurationOptionsImpl<ConfigOption*>(configurationOptions_,
+                                                 pathPrefix);
 }
 
 // ____________________________________________________________________________
-auto ConfigManager::configurationOptions() const {
-  return configurationOptionsImpl(configurationOptions_);
+std::vector<std::pair<std::string, const ConfigOption*>>
+ConfigManager::configurationOptions(std::string_view pathPrefix) const {
+  return configurationOptionsImpl<const ConfigOption*>(configurationOptions_,
+                                                       pathPrefix);
```

Comment:
Why do those return `ConfigOption*` and not `ConfigOption&`? You could then get rid of some dereference lambdas etc.

Reply:
I mainly prefer pointer, because their handling is way easier. But I made the wanted change in [c537d63](https://github.com/ad-freiburg/qlever/pull/1071/commits/c537d63d6c0765a05259e0aa92a263c8bd61319b).